File: js/src/orchestration/pending.ts
/**
 * Pending state management — tracks confirmations awaiting user response.
 */

export interface PendingState {
  type: string;
  payload: Record<string, unknown>;
  createdAt: number;
}

export type ConfirmationDisposition = "confirm" | "reject" | "override";

export interface BasePendingStateManager {
  getPending(sessionId: string): Promise<PendingState | null>;
  setPending(sessionId: string, state: PendingState): Promise<void>;
  clearPending(sessionId: string): Promise<void>;
}

/**
 * InMemoryPendingStateManager — simple in-memory Map implementation.
 */
export class InMemoryPendingStateManager implements BasePendingStateManager {
  private store = new Map<string, PendingState>();

  async getPending(sessionId: string): Promise<PendingState | null> {
    return this.store.get(sessionId) ?? null;
  }

  async setPending(sessionId: string, state: PendingState): Promise<void> {
    this.store.set(sessionId, state);
  }

  async clearPending(sessionId: string): Promise<void> {
    this.store.delete(sessionId);
  }
}

const CONFIRM_PATTERN = /^(好|確認|對|yes|y|ok|好的|是|沒問題|sure|confirm)$/i;
const REJECT_PATTERN = /^(不要|取消|no|n|不|cancel|否|算了|不行|reject)$/i;

/**
 * Default implementation to classify a user message's disposition toward a pending confirmation.
 */
export function classifyConfirmationDisposition(
  message: string,
): ConfirmationDisposition {
  const trimmed = message.trim();
  if (CONFIRM_PATTERN.test(trimmed)) return "confirm";
  if (REJECT_PATTERN.test(trimmed)) return "reject";
  return "override";
}

File: naru_agent/orchestration/pending.py
"""Pending state management — tracks confirmations awaiting user response."""

from __future__ import annotations

import re
import time
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Literal

ConfirmationDisposition = Literal["confirm", "reject", "override"]

_CONFIRM_PATTERN = re.compile(
    r"^(好|確認|對|yes|y|ok|好的|是|沒問題|sure|confirm)$",
    re.IGNORECASE,
)
_REJECT_PATTERN = re.compile(
    r"^(不要|取消|no|n|不|cancel|否|算了|不行|reject)$",
    re.IGNORECASE,
)


def classify_confirmation_disposition(message: str) -> ConfirmationDisposition:
    """Default implementation to classify a user message's disposition toward a pending confirmation."""
    trimmed = message.strip()
    if _CONFIRM_PATTERN.match(trimmed):
        return "confirm"
    if _REJECT_PATTERN.match(trimmed):
        return "reject"
    return "override"


@dataclass
class PendingState:
    """Represents a pending action awaiting user confirmation."""

    type: str  # noqa: A003
    payload: dict[str, Any] = field(default_factory=dict)
    created_at: float = field(default_factory=time.time)


class BasePendingStateManager(ABC):
    """Abstract base for pending state storage backends."""

    @abstractmethod
    def get_pending(self, session_id: str) -> PendingState | None:
        ...

    @abstractmethod
    def set_pending(self, session_id: str, state: PendingState) -> None:
        ...

    @abstractmethod
    def clear_pending(self, session_id: str) -> None:
        ...


class InMemoryPendingStateManager(BasePendingStateManager):
    """Simple in-memory dict implementation of BasePendingStateManager."""

    def __init__(self) -> None:
        self._store: dict[str, PendingState] = {}

    def get_pending(self, session_id: str) -> PendingState | None:
        return self._store.get(session_id)

    def set_pending(self, session_id: str, state: PendingState) -> None:
        self._store[session_id] = state

    def clear_pending(self, session_id: str) -> None:
        self._store.pop(session_id, None)
